src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -513,15 +555,25 @@ void recordWriteAfterQueryExecution() throws SQLException {
         }
     }
 
-    public Statement getReadStatement(String sql) {
+    public Statement getReadStatement(RouteDecisionBuilder decisionBuilder) {
+        if (connectionProvider.getState().equals(MAIN)) {
+            decisionBuilder.reason(MAIN_CONNECTION_REUSE);
+            connectionProvider.getStateDecision().ifPresent(decisionBuilder::cause);
```

Comment:
New reason for deleting builder: it's a mutable object being passed around. With just an immutable `RouteDecision`, we'd have:
```suggestion
            RouteDecision chainedDecision = connectionProvider.getStateDecision().map(decision::withCause).orElse(decision);
```

Reply:
What do you mean by deleting the builder? The main reason for the builder is not to reveal unnecessary API like witchCause with the RouteDecission. There's no use case for this API for consumers of the library.

Reply:
We can make the builder immutable if it's the concern.

Reply:
> The main reason for the builder is not to reveal unnecessary API like witchCause with the RouteDecission.

I didn't realize that. Thanks

Reply:
Exposing `withCause` wouldn't hurt us, because the object is still immutable and it doesn't leak any new data. The lib consumers can reimplement that trivially. Hiding it doesn't help maintainability that much. Without the builder, the source of the impl could be easier to read.
It's a preference thing, not a must-have, not required for the merge.

Reply:
https://github.com/wyrzyk/db-replica/pull/6